### PR TITLE
Introduce debug subcommand

### LINKED ISSUE #660: Feature: roll debug tools verify_invoke.py and container_debug.py into a subcommand debug
**Describe the Feature**
I have been using `tern/tools/verify_invoke.py` and `tern/tools/container_debug.py` to test the scripts that get added to `base.yml` and `snippets.yml` or even just for looking around a container filesystem at a certain layer. However, they have been a pain to maintain as the project keeps growing and moving forward. Hence I propose to roll these tools into Tern as a new subcommand called `debug`.

**Use Cases**
"Debugging" container images could work well as a feature for developers to tinker with container image filesystems, although I see this more as a tool for folks working on Tern itself.

**Implementation Changes**
This would just be using functionality in `verify_invoke.py` and `container_debug.py` under a subcommand called `debug` and removing the scripts from the `tools` folder.

### PR DESCRIPTION
**Pull request by nishakm**
Fixes #660

The "debug" option includes the ability to test out
script in base.yml and to inspect the filesystem at a given
layer index.

To inspect a layer's manifested filesystem run:
`tern debug -i <image:tag> --step`
This will set up an interactive session where the user will be
asked for a layer to inspect. Once chosen, Tern will mount the
layers and provide instructions to inspect the filesystem or or
chroot into it.
To recover the filesystem or clean up run:
`tern debug --recover`

To run a script from base.yml run:
`tern debug -i <image:tag> --keys base <package_manager>
<names/version/etc>`
keys are the keys in base.yml that point to a specific script to
verify.

### CODE REVIEW

**Comment by rnjudge** [comment]
Do we want to put an example of how to use the `--step` debug subcommand in the documentation?

File: tern/analyze/default/debug/run.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+            mount_container_image(full_image, args.driver)
+            # invoke commands in chroot
+            invoke_script(args)
+            # undo the mounts
+            rootfs.undo_mount()
+            rootfs.unmount_rootfs()
+            # cleanup
+            rootfs.clean_up()
+    else:
+        print("Something when wrong in loading the image")
```
Comment (rnjudge) [inline tern/analyze/default/debug/run.py]: Small nit: something `went` wrong.

**Comment by nishakm** [comment]
> Do we want to put an example of how to use the `--step` debug subcommand in the documentation?

Sure, I can include that with the final commit. Should it go in CONTRIBUTING.md since it's a tern development specific feature?

**Comment by rnjudge** [comment]
1) I got an error when testing the `--keys` option on a golang image with lots of layers. I did not hit this with a single-layer photon image.

```
# tern debug -i goabhay:latest --keys base go names --shell "/usr/bin/sh"
.
.
.
Traceback (most recent call last):
  File "/home/rjudge/tt/bin/tern", line 10, in <module>
    sys.exit(main())
  File "/home/rjudge/tt/nisha/tern/tern/__main__.py", line 234, in main
    do_main(args)
  File "/home/rjudge/tt/nisha/tern/tern/__main__.py", line 98, in do_main
    derun.execute_debug(args)
  File "/home/rjudge/tt/nisha/tern/tern/analyze/default/debug/run.py", line 188, in execute_debug
    execute_invoke(full_image, args)
  File "/home/rjudge/tt/nisha/tern/tern/analyze/default/debug/run.py", line 126, in execute_invoke
    mount_container_image(image_obj, args.driver)
  File "/home/rjudge/tt/nisha/tern/tern/analyze/default/debug/run.py", line 44, in mount_container_image
    image_obj, len(image_obj.layers), driver)
  File "/home/rjudge/tt/nisha/tern/tern/analyze/default/container/multi_layer.py", line 32, in mount_overlay_fs
    tar_layers.append(image_obj.layers[index].tar_file)
IndexError: list index out of range
```

2) When running the keys option on a photon single-layer image, the debug output information is somewhat hidden in the terminal output like so:
```
2020-12-10 13:08:07,419 - DEBUG - rootfs - Running command: sudo mount -o bind /dev /home/rjudge/.tern/temp/mergedir/dev
2020-12-10 13:08:07,474 - DEBUG - rootfs - Running command: sudo cp /etc/resolv.conf /home/rjudge/.tern/temp/mergedir/etc/resolv.conf
2020-12-10 13:08:07,499 - DEBUG - rootfs - Running command: sudo unshare -pf --mount-proc=/home/rjudge/.tern/temp/mergedir/proc chroot /home/rjudge/.tern/temp/mergedir /usr/bin/sh -c tdnf check-update > /dev/null && tdnf list installed | cut -f1 -d"."
Output list: bash bzip2-libs ca-certificates ca-certificates-pki curl curl-libs e2fsprogs-libs elfutils-libelf expat expat-libs filesystem glibc krb5 libcap libdb libgcc libmetalink libsolv libssh2 lua ncurses-libs nspr nss-libs openssl photon-release photon-repos popt readline rpm-libs sqlite-libs tdnf tdnf-cli-libs toybox xz-libs zlib
Error messages: 
Number of elements: 35
2020-12-10 13:08:58,477 - DEBUG - rootfs - Running command: sudo umount /home/rjudge/.tern/temp/mergedir/proc
2020-12-10 13:08:58,519 - DEBUG - rootfs - Running command: sudo umount /home/rjudge/.tern/temp/mergedir/sys

```
 do we want to find a way to separate the debug output from the terminal activity output? This might be out of scope for this PR, and if so I can open another issue for it.

3) The step option is nice! Only one small nit that I think needs to change for clarity. When I run `tern debug -i goabhay:latest --step`, I am told to pick a layer, so I do, and the directions I get are:
```
Run 'cd /home/rjudge/.tern/temp/mergedir && sudo chroot . /bin/sh' to look around
Run 'cd -' to go back and 'tern debug --recover' to clean up
```

However, `cd -` without first exiting the chroot environment won't work. So there should be a step to first run `exit` prior to running `cd -`.


3) I  hit the following error when running `tern debug --recover` after doing a debug operation (i.e. `step` or `keys`)on an image that has multiple layers. If I run a debug operation on a single later image, the `--recover` option works properly for me. 
```  (tt) rjudge:mergedir$ tern debug --recover
2020-12-10 13:04:05,983 - DEBUG - __main__ - Starting...
2020-12-10 13:04:05,984 - DEBUG - prep - Setting up...
2020-12-10 13:04:06,127 - DEBUG - rootfs - Running command: sudo umount /home/rjudge/.tern/temp/mergedir/proc
2020-12-10 13:04:06,170 - DEBUG - rootfs - Running command: sudo umount /home/rjudge/.tern/temp/mergedir/sys
2020-12-10 13:04:06,214 - DEBUG - rootfs - Running command: sudo umount /home/rjudge/.tern/temp/mergedir/dev
2020-12-10 13:04:06,261 - DEBUG - rootfs - Running command: sudo umount -rl /home/rjudge/.tern/temp/mergedir
2020-12-10 13:04:06,321 - DEBUG - rootfs - Running command: sudo rm -rf /home/rjudge/.tern/temp/mergedir
2020-12-10 13:04:06,347 - DEBUG - rootfs - Running command: sudo rm -rf /home/rjudge/.tern/temp/workdir
2020-12-10 13:04:06,373 - DEBUG - prep - Tearing down...
Traceback (most recent call last):
  File "/home/rjudge/tt/bin/tern", line 10, in <module>
    sys.exit(main())
  File "/home/rjudge/tt/nisha/tern/tern/__main__.py", line 234, in main
    do_main(args)
  File "/home/rjudge/tt/nisha/tern/tern/__main__.py", line 100, in do_main
    prep.teardown(args.keep_wd)
  File "/home/rjudge/tt/nisha/tern/tern/prep.py", line 44, in teardown
    clean_working_dir()
  File "/home/rjudge/tt/nisha/tern/tern/prep.py", line 62, in clean_working_dir
    shutil.rmtree(path)
  File "/usr/lib/python3.6/shutil.py", line 486, in rmtree
    _rmtree_safe_fd(fd, path, onerror)
  File "/usr/lib/python3.6/shutil.py", line 424, in _rmtree_safe_fd
    _rmtree_safe_fd(dirfd, fullname, onerror)
  File "/usr/lib/python3.6/shutil.py", line 424, in _rmtree_safe_fd
    _rmtree_safe_fd(dirfd, fullname, onerror)
  File "/usr/lib/python3.6/shutil.py", line 424, in _rmtree_safe_fd
    _rmtree_safe_fd(dirfd, fullname, onerror)
  [Previous line repeated 4 more times]
  File "/usr/lib/python3.6/shutil.py", line 444, in _rmtree_safe_fd
    onerror(os.unlink, fullname, sys.exc_info())
  File "/usr/lib/python3.6/shutil.py", line 442, in _rmtree_safe_fd
    os.unlink(name, dir_fd=topfd)
PermissionError: [Errno 13] Permission denied: 'TODO'

```

**Comment by nishakm** [comment]
> > Do we want to put an example of how to use the `--step` debug subcommand in the documentation?
> 
> Sure, I can include that with the final commit. Should it go in CONTRIBUTING.md since it's a tern development specific feature?

@rnjudge any feedback on this?

**Comment by rnjudge** [comment]
> > Do we want to put an example of how to use the `--step` debug subcommand in the documentation?
> 
> Sure, I can include that with the final commit. Should it go in CONTRIBUTING.md since it's a tern development specific feature?

@nishakm I think it might be better as a subsection in adding-to-the-command-library.md? The other debug functionality is there and the step functionality is helpful if someone is trying to add to the command library but not getting the output they expect from the `debug --keys` output. What do you think?

**Comment by nishakm** [comment]
> > > Do we want to put an example of how to use the `--step` debug subcommand in the documentation?
> > 
> > 
> > Sure, I can include that with the final commit. Should it go in CONTRIBUTING.md since it's a tern development specific feature?
> 
> @nishakm I think it might be better as a subsection in adding-to-the-command-library.md? The other debug functionality is there and the step functionality is helpful if someone is trying to add to the command library but not getting the output they expect from the `debug --keys` output. What do you think?

I will put some documentation in there for now.

**Comment by nishakm** [comment]
>     1. When running the keys option on a photon single-layer image, the debug output information is somewhat hidden in the terminal output like so:
> 
> 
> ```
> 2020-12-10 13:08:07,419 - DEBUG - rootfs - Running command: sudo mount -o bind /dev /home/rjudge/.tern/temp/mergedir/dev
> 2020-12-10 13:08:07,474 - DEBUG - rootfs - Running command: sudo cp /etc/resolv.conf /home/rjudge/.tern/temp/mergedir/etc/resolv.conf
> 2020-12-10 13:08:07,499 - DEBUG - rootfs - Running command: sudo unshare -pf --mount-proc=/home/rjudge/.tern/temp/mergedir/proc chroot /home/rjudge/.tern/temp/mergedir /usr/bin/sh -c tdnf check-update > /dev/null && tdnf list installed | cut -f1 -d"."
> Output list: bash bzip2-libs ca-certificates ca-certificates-pki curl curl-libs e2fsprogs-libs elfutils-libelf expat expat-libs filesystem glibc krb5 libcap libdb libgcc libmetalink libsolv libssh2 lua ncurses-libs nspr nss-libs openssl photon-release photon-repos popt readline rpm-libs sqlite-libs tdnf tdnf-cli-libs toybox xz-libs zlib
> Error messages: 
> Number of elements: 35
> 2020-12-10 13:08:58,477 - DEBUG - rootfs - Running command: sudo umount /home/rjudge/.tern/temp/mergedir/proc
> 2020-12-10 13:08:58,519 - DEBUG - rootfs - Running command: sudo umount /home/rjudge/.tern/temp/mergedir/sys
> ```
> 
> do we want to find a way to separate the debug output from the terminal activity output? This might be out of scope for this PR, and if so I can open another issue for it.
> 

I've made it look like this:
```
2020-12-11 13:48:30,831 - DEBUG - rootfs - Running command: sudo unshare -pf --mount-proc=/home/nisha/.tern/temp/mergedir/proc chroot /home/nisha/.tern/temp/mergedir /usr/bin/sh -c go list -m all | tail -n +2 | cut -d ' ' -f 1
2020-12-11 13:48:30,937 - ERROR - rootfs - Command failed. go list -m: not using modules

2020-12-11 13:48:30,938 - WARNING - rootfs - Error executing command in chroot
2020-12-11 13:48:30,938 - WARNING - collect - Error executing snippets: Command 'go list -m all | tail -n +2 | cut -d ' ' -f 1' returned non-zero exit status 1.

*************************************************************
          Command Library Script Verification mode           
*************************************************************

Output list: 
Error messages: go list -m: not using modules

Number of elements: 0

2020-12-11 13:48:30,939 - DEBUG - rootfs - Running command: sudo umount /home/nisha/.tern/temp/mergedir/proc
...
```

>     1. The step option is nice! Only one small nit that I think needs to change for clarity. When I run `tern debug -i goabhay:latest --step`, I am told to pick a layer, so I do, and the directions I get are:
> 
> 
> ```
> Run 'cd /home/rjudge/.tern/temp/mergedir && sudo chroot . /bin/sh' to look around
> Run 'cd -' to go back and 'tern debug --recover' to clean up
> ```
> 
> However, `cd -` without first exiting the chroot environment won't work. So there should be a step to first run `exit` prior to running `cd -`.
> 

I've changed the messages this way:
```
Run 'cd /home/nisha/.tern/temp/mergedir && sudo chroot . /bin/sh' to look around

After exiting from your session, run 'cd -' to go back and 'tern debug --recover' to clean up.
```

**Review by rnjudge — APPROVED** [review]
The errors I was seeing are gone. This looks good! Merging now.

**Pull request merged**